File: src/main/java/com/now/nowbot/model/JsonData/ActivityEvent.java
package com.now.nowbot.model.JsonData;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.now.nowbot.model.enums.OsuMode;

import java.time.OffsetDateTime;

@JsonIgnoreProperties(ignoreUnknown = true)
@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
public class ActivityEvent {
    public record Beatmap(String title, String url) {
    }

    public record BeatmapSet(String title, String url) {
    }


    public record User(String name, String url, String previousUsername) {
    }

    public enum EventType {
        achievement("achievement", "user"),
        beatmapPlaycount("beatmap", "count"),
        beatmapsetApprove("approval", "beatmapset", "user"),
        beatmapsetDelete("beatmapset"),
        beatmapsetRevive("beatmapset", "user"),
        beatmapsetUpdate("beatmapset", "user"),
        beatmapsetUpload("beatmapset", "user"),

        rank("scoreRank", "rank", "mode", "beatmap", "user"),
        rankLost("mode", "beatmap", "user"),
        userSupportAgain("user"),
        userSupportFirst("user"),
        userSupportGift("user"),
        usernameChange("user"),
        ;
        final String[] field;

        EventType(String... f) {
            field = f;
        }
    }

    @JsonProperty("created_at")
//    @JsonFormat(locale = "zh", timezone = "GMT+8", pattern = "yyyy-MM-dd'T'HH:mm:ss.SSSZ")
    OffsetDateTime createdAt;
    @JsonProperty("id")
    Long id;

    @JsonIgnore
    EventType type;

    @JsonProperty("type")
    void setType(String type) {
        this.type = EventType.valueOf(type);
    }
    /*****************************下面属性根据不同 type 进行变动, 只包含 enum 中括号内的属性************************************/
    @JsonProperty("count")
    Integer count;
    @JsonProperty("approval")
    Integer approval;
    @JsonProperty("scoreRank")
    String scoreRank;
    @JsonProperty("rank")
    Integer rank;
    @JsonProperty("mode")
    String mode;

    @JsonProperty("achievement")
    String achievement;

    @JsonProperty("user")
    User user;
    @JsonProperty("beatmap")
    Beatmap beatmap;
    @JsonProperty("beatmapset")
    BeatmapSet beatmapSet;

    public OffsetDateTime getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(OffsetDateTime createdAt) {
        this.createdAt = createdAt;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public Integer getCount() {
        return count;
    }

    public void setCount(Integer count) {
        this.count = count;
    }

    public Integer getApproval() {
        return approval;
    }

    public void setApproval(Integer approval) {
        this.approval = approval;
    }

    public String getScoreRank() {
        return scoreRank;
    }

    public void setScoreRank(String scoreRank) {
        this.scoreRank = scoreRank;
    }

    public Integer getRank() {
        return rank;
    }

    public void setRank(Integer rank) {
        this.rank = rank;
    }

    public String getAchievement() {
        return achievement;
    }

    public void setAchievement(String achievement) {
        this.achievement = achievement;
    }

    public User getUser() {
        return user;
    }

    public void setUser(User user) {
        this.user = user;
    }

    public Beatmap getBeatmap() {
        return beatmap;
    }

    public void setBeatmap(Beatmap beatmap) {
        this.beatmap = beatmap;
    }

    public BeatmapSet getBeatmapSet() {
        return beatmapSet;
    }

    public void setBeatmapSet(BeatmapSet beatmapSet) {
        this.beatmapSet = beatmapSet;
    }

    public EventType getType() {
        return type;
    }

    public OsuMode getMode() {
        return OsuMode.getMode(mode);
    }

    public boolean isTypeMapping() {
        return (type == EventType.beatmapsetApprove ||
                type == EventType.beatmapsetDelete ||
                type == EventType.beatmapsetRevive ||
                type == EventType.beatmapsetUpdate ||
                type == EventType.beatmapsetUpload
        );
    }

    @Override
    public String toString() {
        return STR."ActivityEvent{createdAt=\{createdAt}, id=\{id}, count=\{count}, approval=\{approval}, scoreRank='\{scoreRank}\{'\''}, rank=\{rank}, mode='\{mode}\{'\''}, achievement='\{achievement}\{'\''}, type=\{type}, user=\{user}, beatmap=\{beatmap}, beatmapSet=\{beatmapSet}\{'}'}";
    }
}

File: src/main/java/com/now/nowbot/model/osu/ActivityEvent.kt
package com.now.nowbot.model.osu

import com.fasterxml.jackson.annotation.JsonAutoDetect
import com.fasterxml.jackson.annotation.JsonIgnore
import tools.jackson.databind.PropertyNamingStrategies
import tools.jackson.databind.annotation.JsonNaming
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonProperty
import java.time.OffsetDateTime

@JsonIgnoreProperties(ignoreUnknown = true)
@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy::class)
class ActivityEvent {
    data class Achievement(
        @field:JsonProperty("icon_url") val url: String,
        @field:JsonProperty("id") val achievementID: Int,
        val name: String,
        val grouping: String,
        val ordering: String,
        val slug: String,
        val description: String,
        val mode: String?,
        val instructions: String?
    )

    data class EventBeatmap(
        val title: String,
        @field:JsonProperty("url")
        val url: String?
    ) {
        @get:JsonProperty("id")
        val id: Long
            get() = url?.substringBefore("?")?.split("/")?.lastOrNull()?.toLongOrNull() ?: 0L
    }

    data class EventBeatmapset(
        val title: String,

        @field:JsonProperty("url")
        val url: String?
    ) {
        @get:JsonProperty("id")
        val id: Long
            get() = url?.substringBefore("?")?.split("/")?.lastOrNull()?.toLongOrNull() ?: 0L
    }

    data class EventUser(
        @field:JsonProperty("username")
        val name: String,

        @field:JsonProperty("url")
        val url: String?,

        @field:JsonProperty("previous_username")
        val previousUsername: String?
    ) {
        @get:JsonProperty("id")
        val id: Long
            get() = url?.split("/")?.lastOrNull()?.toLongOrNull() ?: 0L
    }

    enum class EventType(vararg f: String) {
        Achievement("achievement", "user"),
        BeatmapPlaycount("beatmap", "count"),
        BeatmapsetApprove("approval", "beatmapset", "user"),
        BeatmapsetDelete("beatmapset"),
        BeatmapsetRevive("beatmapset", "user"),
        BeatmapsetUpdate("beatmapset", "user"),
        BeatmapsetUpload("beatmapset", "user"),

        Rank("scoreRank", "rank", "mode", "beatmap", "user"),
        RankLost("mode", "beatmap", "user"),
        UserSupportAgain("user"),
        UserSupportFirst("user"),
        UserSupportGift("user"),
        UsernameChange("user"),
        ;

        val field: Array<String> = f.toList().toTypedArray()

        companion object {
            fun getEventType(string: String?): EventType? {
                if (string.isNullOrBlank()) return null

                EventType.entries.forEach {
                    if (it.name.equals(string.trim(), true)) {
                        return it
                    }
                }

                return null
            }
        }
    }

    //@JsonFormat(locale = "zh", timezone = "GMT+8", pattern = "yyyy-MM-dd'T'HH:mm:ss.SSSZ")
    @JsonProperty("created_at")
    var createdAt: OffsetDateTime? = null

    @JsonProperty("id")
    var eventID: Long = 0

    @get:JsonIgnore
    var type: EventType? = null

    @JsonProperty("type") fun setType(type: String?) {
        this.type = EventType.getEventType(type)
    }

    /*****************************下面属性根据不同 type 进行变动, 只包含 enum 中括号内的属性 */
    @JsonProperty("count")
    var count: Int? = null

    @JsonProperty("approval")
    var approval: String? = null

    @JsonProperty("scoreRank")
    var scoreRank: String? = null

    @JsonProperty("rank")
    var rank: Int? = null

    @JsonProperty("mode")
    var mode: String? = null

    @JsonProperty("achievement")
    var achievement: Achievement? = null

    @JsonProperty("user")
    var user: EventUser? = null

    @JsonProperty("beatmap")
    var beatmap: EventBeatmap? = null

    @JsonProperty("beatmapset")
    var beatmapSet: EventBeatmapset? = null

    val isMapping: Boolean
        get() = type in MAPPING_EVENTS

    // 自己记录，List.squash 后记录结果
    @JsonProperty("repeated")
    var repeated: Int = 1

    override fun equals(other: Any?): Boolean {
        if (other is ActivityEvent) {
            return this.type == other.type && (beatmapSet?.url == other.beatmapSet?.url)
        }
        return false
    }

    override fun hashCode(): Int {
        return (this.type!!.name + (beatmapSet?.url ?: "")).hashCode()
    }

    override fun toString(): String {
        return "ActivityEvent(createdAt=$createdAt, eventID=$eventID, type=$type, count=$count, approval=$approval, scoreRank=$scoreRank, rank=$rank, mode=$mode, achievement=$achievement, user=$user, beatmap=$beatmap, beatmapSet=$beatmapSet, isMapping=$isMapping)"
    }

    companion object {
        private val MAPPING_EVENTS = setOf(
            EventType.BeatmapsetApprove,
            EventType.BeatmapsetDelete,
            EventType.BeatmapsetRevive,
            EventType.BeatmapsetUpdate,
            EventType.BeatmapsetUpload
        )

        fun List<ActivityEvent>.filterIsMapping(): List<ActivityEvent> {
            return this.filter { it.isMapping }
        }

        /**
         * 合并日期相同、谱面相同、类型相同的 event
         */
        fun List<ActivityEvent>.squash(): List<ActivityEvent> {
            if (this.isEmpty()) return emptyList()

            data class EventGroupKey(
                val date: java.time.LocalDate?,
                val type: EventType?,
                val beatmapID: Long?
            )

            val groups = linkedMapOf<EventGroupKey, ActivityEvent>()

            for (current in this) {
                val key = EventGroupKey(
                    date = current.createdAt?.toLocalDate(),
                    type = current.type,
                    beatmapID = current.beatmap?.id
                )

                val firstSaved = groups[key]
                if (firstSaved != null) {
                    firstSaved.repeated++
                } else {
                    groups[key] = current
                }
            }

            return groups.values.toList()
        }
    }
}
